src/vm/database/clarity_db.rs

```diff
@@ -680,6 +719,22 @@ impl<'a> ClarityDatabase<'a> {
             .expect("Failed to get block data.")
     }
 
+    /// Returns the header hash of the burnchain block at `burnchain_block_height`
+    /// using the "canonical sortition" (see SortitionDB::get_canonical_sortition_tip).
+    ///
+    /// Returns None if `height` is negative or greater than the canonical burn
+    /// chain height.
+    ///
+    /// # Panics
+    /// Panics if the underlying database is not available.
+    pub fn get_burnchain_block_header_hash_for_burnchain_height(
+        &mut self,
+        burnchain_block_height: u32,
+    ) -> Option<BurnchainHeaderHash> {
+        self.burn_state_db
+            .get_burn_header_hash_using_canonical_sortition(burnchain_block_height)
```

Comment:
I don't think that it is sufficient to use the canonical sortition for looking up the burnchain block at a given burnchain height here. The way that these lookups are performed is that the provided "sortition" defines the burnchain fork that the lookup is performed from. The canonical sortition _may not_ be the same fork that the currently-being-evaluated Stacks block was contained within (I believe that with the way the ChainsCoordinator currently works, this _is_ the case, but it's not a guarantee that the ChainsCoordinator explicitly provides).

For example, the burnchain could look like this:

```
A -> B -> C -> D
 \_> B' -> C'
```

Where the currently-being-evaluated Stacks block was elected in `C'`, but the canonical sortition is `D`. We'd want `get-burn-block-info u1` to return `B'` in this case and _not_ `B`.

The second important thing here is that the height supplied to `get-burn-block-info` _must_ be strictly less than the current burn block height. This is because miners have to evaluate the block _before_ the burn block is mined, so they could not _know_ that `(get-burn-block-info u2)` should return `C'`. Basically, if the supplied height is `>= current_burn_block_height` it should return `none`.

Reply:
Beat me to it.

You're going to need to track the `StacksBlockId` of the chain tip at which this code is running, and use *that* to look up the associated burnchain block hash.  Then, if the VM is in the process of building block `C'` off of `B'`, then this method would be able to report the hash of `B'` as the hash of the block at height 2.  Correspondingly, if the VM is in the process of building block `C` off of `B`, then this method would be able to report the hash of `B` as the hash of the block at height 2.